query/src/main/java/io/keen/client/java/CachedDatasetRequest.java

```diff
@@ -0,0 +1,138 @@
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.*;
+
+import static java.util.Collections.emptyMap;
+
+abstract class CachedDatasetRequest extends PersistentAnalysis {
+
+    private CachedDatasetRequest(String httpMethod, boolean needsMasterKey, String datasetName) {
+        super(httpMethod, needsMasterKey, datasetName, null);
+    }
+
+    static KeenQueryRequest definitionRequest(String datasetName) {
+        return new CachedDatasetRequest(HttpMethods.GET, false, datasetName) {
+
+            @Override
+            URL getRequestURL(RequestUrlBuilder urlBuilder, String projectId) throws KeenQueryClientException {
+                return urlBuilder.getDatasetsUrl(projectId, getResourceName(), false, Collections.<String, Object>emptyMap());
+            }
+        };
+    }
+
+    static KeenQueryRequest resultsRequest(final DatasetDefinition datasetDefinition, final String indexBy, final Timeframe timeframe) {
+        return new CachedDatasetRequest(HttpMethods.GET, false, datasetDefinition.getDatasetName()) {
+            private final ObjectMapper objectMapper = new ObjectMapper()
+                    .configure(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS, false);
+
+            @Override
+            URL getRequestURL(RequestUrlBuilder urlBuilder, String projectId) throws KeenQueryClientException {
+                try {
+                    return urlBuilder.getDatasetsUrl(projectId, datasetDefinition.getDatasetName(), true, queryArgs());
+                } catch (IOException e) {
+                    throw new KeenQueryClientException(e);
+                }
+            }
+
+            @Override
+            boolean retrievingResults() {
+                return true;
+            }
+
+            @Override
+            boolean groupedResponseExpected() {
+                return getGroupByParams() != null && !getGroupByParams().isEmpty();
+            }
+
+            @Override
+            Collection<String> getGroupByParams() {
+                return datasetDefinition.getQuery().getGroupBy();
+            }
+
+            private Map<String, Object> queryArgs() throws IOException {
+                if (indexBy == null) {
+                    throw new IllegalArgumentException("index_by is required");
+                }
+                if (timeframe == null) {
+                    throw new IllegalArgumentException("timeframe is required");
+                }
+
+                Object timeframeAsArgs = timeframe.constructTimeframeArgs().get(KeenQueryConstants.TIMEFRAME);
+
+                HashMap<String, Object> queryArgs = new HashMap<String, Object>();
+                queryArgs.put(KeenQueryConstants.INDEX_BY, indexBy);
+                queryArgs.put(KeenQueryConstants.TIMEFRAME, objectMapper.writeValueAsString(timeframeAsArgs));
+                return queryArgs;
+            }
+        };
+    }
+
+    static KeenQueryRequest definitionsByProjectRequest(final Integer limit, final String afterName) {
```

Comment:
Is it intentional that we're boxing `limit` here? Is that because it goes into the `Map` as an `Object` anyway later at url construction time? It seems like it could just as well be boxed at the last moment if possible.

Reply:
It is intentional. Both `limit` and `afterName` are optional params. On the API level we expose overloaded versions of getDefinitionsByProject

```
    List<DatasetDefinition> getDefinitionsByProject() throws IOException;

    List<DatasetDefinition> getDefinitionsByProject(Integer limit, String afterName) throws IOException;

```